src/persist/src/indexed/future.rs

```diff
@@ -91,8 +91,21 @@ impl<K: Data, V: Data> BlobFuture<K, V> {
                 batch.desc
             )));
         }
-        // TODO: Assert that nothing in the batch is before self.ts_lower. That
-        // would indicate a logic error by the user of this BlobFuture.
+        if cfg!(any(debug, test)) {
+            // Batches being appended to this future come from data being
+            // drained out of the buffer. Indexed should have prevented this
+            // write to the buffer, so this should never happen. Hopefully any
+            // regressions in maintaining this invariant will be caught by this
+            // debug/test check.
+            for (_, _, ts, _) in batch.updates.iter() {
+                if !self.ts_lower.less_equal(ts) {
```

Comment:
I thought that since ts_lower was a closed lower bound updates at ts_lower would be allowed.

Like, taking a concrete example because thinking about this abstractly lets say that initially future logically contains records at times [0, 10) and trace is empty (contains records at [0, 0) ).

then a user does `seal(4)`

then my understanding is that future contains records from [4, 10) and trace contains records from [0, 4) logically.

at this point my understanding is that new records at 4 would still be allowed. does that sound correct?

Reply:
This check is actually correct (I think). If `ts_lower == ts`, then `less_equal` returns true and `!less_equal` evaluates as false, so the `if` block which returns an error is skipped (aka `ts_lower == ts` is allowed). I've added a test case

Reply:
oh duh right yes that makes total sense!